### DISCUSSION TITLE
[Android] How to call native extension function in main thread?

### DISCUSSION THREAD

**Discussion by itlancer** [airsdk/Adobe-Runtime-Support#2787] [Q&A] [answered]
How to call Android native extension function in main thread if application use `<runtimeInBackgroundThread>true</runtimeInBackgroundThread>` in manifest?

If I set up `<runtimeInBackgroundThread>true</runtimeInBackgroundThread>` then some of my Android native extension functions failed with errors:
```
android.view.ViewRootImpl$CalledFromWrongThreadException: Only the original thread that created a view hierarchy can touch its views.
	at android.view.ViewRootImpl.checkThread(ViewRootImpl.java:9629)
	at android.view.ViewRootImpl.requestLayout(ViewRootImpl.java:1919)
	at android.view.View.requestLayout(View.java:26296)
	at android.view.ViewGroup.addView(ViewGroup.java:5075)
	at android.view.ViewGroup.addView(ViewGroup.java:5017)
	at android.view.ViewGroup.addView(ViewGroup.java:4989)
```

So I need to call them in main application thread. I know that new method `FREContext::callFunctionInMainThread()` was created but cannot figure out how to use it properly.
I've tried to create one Java native extension function:
```
package com.test;

import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.adobe.fre.FREFunction;

public class ANEFunction implements FREFunction {

	@Override
	public FREObject call(FREContext freContext, FREObject[] args) {
		freContext.callFunctionInMainThread(new NativeVideoMainUIThreadFunction(), args);

		return null;
	}
}
```
that call another one (with logic) using `FREContext::callFunctionInMainThread()`:
```
package com.test;

import com.adobe.fre.FREArray;
import com.adobe.fre.FREContext;
import com.adobe.fre.FREObject;
import com.adobe.fre.FREFunction;

public class ANEMainUIThreadFunction implements FREFunction {
	
	@Override
	public FREObject call(FREContext freContext, FREObject[] args) {
		try {
			final FREArray array = (FREArray) args[0];
			final String command = array.getObjectAt(0).getAsString();//This line cause com.adobe.fre.FREWrongThreadException
			//... addView()
		} catch (Exception e){
			e.printStackTrace();
		}
		return null;
	}
}
```

But it failed with FREWrongThreadException:
```
W/System.err: com.adobe.fre.FREWrongThreadException
W/System.err:     at com.adobe.fre.FREArray.getObjectAt(Native Method)
W/System.err:     at com.test.ANEMainUIThreadFunction.call(ANEMainUIThreadFunction.java:25)
W/System.err:     at com.adobe.fre.FREContext$1.run(FREContext.java:184)
W/System.err:     at android.os.Handler.handleCallback(Handler.java:942)
W/System.err:     at android.os.Handler.dispatchMessage(Handler.java:99)
W/System.err:     at android.os.Looper.loopOnce(Looper.java:346)
W/System.err:     at android.os.Looper.loop(Looper.java:475)
W/System.err:     at android.app.ActivityThread.main(ActivityThread.java:7950)
W/System.err:     at java.lang.reflect.Method.invoke(Native Method)
W/System.err:     at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
W/System.err:     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:942)
```

**Comment by ajwfrost** [comment, marked as answer]
That `callFunctionInMainThread` function is what we use internally.. I suppose it might be possible for you to call this manually, but not from your own function as this would then get called back again and probably result in some weird deadlock thing.. It's called on the freContext object, the first argument is your FREFunction object and then the arguments to that function's `call` method should be passed. But yes, don't call it from itself, you could perhaps have two FREFunction implementations, one which is passed back to the runtime and then just calls the other one via this mechanism?

-----

Anyway -> we added a workaround for this which will mean that the runtime triggers your FREFunction.call() from the UI thread rather than from the runtime's background thread .. see https://github.com/airsdk/Adobe-Runtime-Support/issues/2326#issuecomment-1498528867 - essentially, during start-up we check for the existence of a file which has the same name as the function.

We should look at a better way of triggering this .. maybe an option on the FREFunction table would be a simpler way of doing it!

----

But alternatively -> if this is an extension for which you have the source code, then the better option would be to wrap your functionality within a 'runOnUiThread' section;
```
freContext.getActivity().runOnUiThread(new Runnable() {
    public void run() {
        Toast.makeText(activity, "Hello, world!", Toast.LENGTH_SHORT).show();
    }
});
```

**Reply by itlancer** [reply]
@ajwfrost
Thank you very much `runOnUiThread()` did the trick for our native extension!

But this https://github.com/airsdk/Adobe-Runtime-Support/issues/2326#issuecomment-1498528867 approach with "marking" functions doesn't work if native function use `FREObject.getAsString()` or something like that.
You will get `FREWrongThreadException` as mentioned: https://github.com/airsdk/Adobe-Runtime-Support/issues/2326#issuecomment-1528854188

**Reply by ajwfrost** [reply]
> But this approach with "marking" functions doesn't work if native function use FREObject.getAsString() or something like that.
You will get FREWrongThreadException

Yes, that's true .. and that `FREWrongThreadException` is not something we'd be able to work around.. 

So I guess if you need arguments and any AS3 interactions, then it needs to be running in the normal runtime thread. At which point, if you also need Android UI/view manipulation, then you need to do the `runOnUiThread` thing. Bit clunky I guess but not sure there's much else we could do here..

thanks
